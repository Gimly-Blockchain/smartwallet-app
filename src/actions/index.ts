import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'
import * as interactionHandlerActions from 'src/actions/interactionHandler'
import * as paymentActions from 'src/actions/payment/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
  interactionHandlerActions,
  paymentActions
}
