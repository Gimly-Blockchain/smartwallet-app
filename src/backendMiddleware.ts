import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { EncryptionLib, EncryptionLibInterface } from 'src/lib/crypto'
import { Storage } from 'src/lib/storage/storage'
import { KeyChain, KeyChainInterface } from 'src/lib/keychain'
import { ConnectionOptions } from 'typeorm/browser'
import {
  createJolocomRegistry,
  JolocomRegistry,
} from 'jolocom-lib/js/registries/jolocomRegistry'
import { IpfsCustomConnector } from './lib/ipfs'
import { jolocomContractsAdapter } from 'jolocom-lib/js/contracts/contractsAdapter'
import { jolocomEthereumResolver } from 'jolocom-lib/js/ethereum/ethereum'
import { jolocomContractsGateway } from 'jolocom-lib/js/contracts/contractsGateway'
import { JolocomLib } from 'jolocom-lib'
import { KeyTypes } from 'jolocom-lib/js/vaultedKeyProvider/types'
import { publicKeyToDID } from 'jolocom-lib/js/utils/crypto'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { generateSecureRandomBytes } from './lib/util';

export enum ErrorCodes {
  NoEntropy = 'NoEntropy',
  NoKeyProvider = 'NoKeyProvider',
  NoWallet = 'NoWallet',
  DecryptionFailed = 'DecryptionFailed',
}

export class BackendError extends Error {
  static codes = ErrorCodes

  constructor(code: ErrorCodes) {
    super(code)
  }
}

export class BackendMiddleware {
  private _identityWallet!: IdentityWallet
  private _keyProvider!: SoftwareKeyProvider
  private _entropyData!: { encryptedEntropy: string; timestamp: number }

  public storageLib: Storage
  public encryptionLib: EncryptionLibInterface
  public keyChainLib: KeyChainInterface
  public registry: JolocomRegistry

  public constructor(config: {
    fuelingEndpoint: string
    typeOrmConfig: ConnectionOptions
  }) {
    this.storageLib = new Storage(config.typeOrmConfig)
    this.encryptionLib = new EncryptionLib()
    this.keyChainLib = new KeyChain()
    this.registry = createJolocomRegistry({
      ipfsConnector: new IpfsCustomConnector({
        host: 'ipfs.jolocom.com',
        port: 443,
        protocol: 'https',
      }),
      ethereumConnector: jolocomEthereumResolver,
      contracts: {
        adapter: jolocomContractsAdapter,
        gateway: jolocomContractsGateway,
      },
    })
  }

  public async initStorage(): Promise<void> {
    await this.storageLib.initConnection()
  }

  public get identityWallet(): IdentityWallet {
    if (this._identityWallet) return this._identityWallet
    throw new BackendError(ErrorCodes.NoWallet)
  }

  public get keyProvider(): SoftwareKeyProvider {
    if (this._keyProvider) return this._keyProvider
    throw new BackendError(ErrorCodes.NoKeyProvider)
  }

  public get entropyData(): BackendMiddleware['_entropyData'] {
    if (this._entropyData) return this._entropyData
    throw new BackendError(ErrorCodes.NoEntropy)
  }

  public async prepareIdentityWallet(): Promise<IdentityWallet> {
    if (this._identityWallet) return this._identityWallet

    const encryptedEntropy = await this.storageLib.get.encryptedSeed()
    if (!encryptedEntropy) throw new BackendError(ErrorCodes.NoEntropy)

    this._entropyData = { encryptedEntropy, timestamp: Date.now() }
    const encryptionPass = await this.keyChainLib.getPassword()

    const decryptedSeed = this.encryptionLib.decryptWithPass({
      cipher: this._entropyData.encryptedEntropy,
      pass: encryptionPass,
    })

    if (!decryptedSeed) {
      throw new BackendError(ErrorCodes.DecryptionFailed)
    }

    // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
    this._keyProvider = JolocomLib.KeyProvider.fromSeed(
      Buffer.from(decryptedSeed, 'hex'),
      encryptionPass,
    )

    const userPubKey = this._keyProvider.getPublicKey({
      derivationPath: KeyTypes.jolocomIdentityKey,
      encryptionPass,
    })

    const didDocument = await this.storageLib.get.didDoc(
      publicKeyToDID(userPubKey),
    )

    if (didDocument) {
      const identity = Identity.fromDidDocument({ didDocument })

      // TODO Simplify constructor
      return (this._identityWallet = new IdentityWallet({
        identity,
        vaultedKeyProvider: this._keyProvider,
        publicKeyMetadata: {
          derivationPath: KeyTypes.jolocomIdentityKey,
          keyId: identity.publicKeySection[0].id,
        },
        contractsAdapter: this.registry.contractsAdapter,
        contractsGateway: this.registry.contractsGateway,
      }))
    } else {
      const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes
      const identityWallet = await this.registry.authenticate(
        this._keyProvider,
        {
          encryptionPass,
          derivationPath,
        },
      )

      await this.storageLib.store.didDoc(identityWallet.didDocument)
      return (this._identityWallet = identityWallet)
    }
  }

  public async setEntropy(encodedEntropy: string): Promise<void> {
    const password = (await generateSecureRandomBytes(32)).toString('base64')
    const encEntropy = this.encryptionLib.encryptWithPass({
      data: encodedEntropy,
      pass: password,
    })
    this._entropyData = { encryptedEntropy: encEntropy, timestamp: Date.now() }

    // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
    this._keyProvider = JolocomLib.KeyProvider.fromSeed(
      Buffer.from(encodedEntropy, 'hex'),
      password,
    )
  }

  public async fuelKeyWithEther(): Promise<void> {
    const password = await this.keyChainLib.getPassword()
    await JolocomLib.util.fuelKeyWithEther(
      this.keyProvider.getPublicKey({
        encryptionPass: password,
        derivationPath: JolocomLib.KeyTypes.ethereumKey,
      }),
    )
  }

  public async createIdentity(): Promise<Identity> {
    const password = await this.keyChainLib.getPassword()
    this._identityWallet = await this.registry.create(
      this.keyProvider,
      password,
    )

    const personaData = {
      did: this._identityWallet.identity.did,
      controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
    }
    await this.storageLib.store.persona(personaData)
    await this.storageLib.store.didDoc(this._identityWallet.didDocument)
    await this.storageLib.store.encryptedSeed(this._entropyData)
    await this.keyChainLib.savePassword(password)

    return this._identityWallet.identity
  }
}
