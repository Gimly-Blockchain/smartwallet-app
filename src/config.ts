import { ConnectionOptions } from 'typeorm/browser'
import { entityList } from './lib/storage/entities'

const typeOrmConf: ConnectionOptions = {
  type: 'react-native',
  database: 'LocalSmartWalletData',
  location: 'default',
  logging: ['error', 'query', 'schema'],
  synchronize: true,
  entities: entityList,
}

const env = process.env['NODE_ENV'] || 'development'
const isDev = env === 'development'
const isTest = env === 'test'

export const skipEntropyCollection = isDev || isTest
export const skipIdentityRegisteration = isDev || isTest

export default {
  fuelingEndpoint: 'https://faucet.jolocom.com/request',
  typeOrmConfig: typeOrmConf,
}
