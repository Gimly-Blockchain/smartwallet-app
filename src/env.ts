const env = process.env['NODE_ENV'] || 'development'
const isDev = env === 'development'
// @ts-ignore unused so far
const isTest = env === 'test'
const isTestE2E = env === 'test-e2e'

export const SKIP_ENTROPY_COLLECTION = isDev || isTest || isTestE2E
export const SKIP_IDENTITY_REGISTRATION = isDev || isTestE2E
