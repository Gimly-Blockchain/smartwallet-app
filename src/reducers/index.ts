import { combineReducers } from 'redux'
import { settingsReducer, SettingsState } from 'src/reducers/settings/'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  settings: settingsReducer,
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly settings: SettingsState
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
