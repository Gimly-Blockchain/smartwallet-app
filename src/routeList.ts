// NOTE: don't use 'const' so that values are useable in both .js and .ts files
export enum routeList {
  AppInit = 'AppInit',

  Landing = 'Landing',
  Entropy = 'Entropy',
  RegistrationProgress = 'RegistrationProgress',
  Loading = 'Loading',

  Home = 'Home',
  Claims = 'Claims',
  Documents = 'Documents',
  Records = 'Records',
  Settings = 'Settings',
  QRCodeScanner = 'QRCodeScanner',

  CredentialDialog = 'CredentialDialog',
  Consent = 'Consent',
  PaymentConsent = 'PaymentConsent',
  AuthenticationConsent = 'AuthenticationConsent',
  ClaimDetails = 'ClaimDetails',
  DocumentDetails = 'DocumentDetails',

  SeedPhrase = 'SeedPhrase',
  RepeatSeedPhrase = 'RepeatSeedPhrase',
  SocialRecovery = 'SocialRecovery',
  ReceivedShards = 'ReceivedShards',

  Exception = 'Exception',
}
