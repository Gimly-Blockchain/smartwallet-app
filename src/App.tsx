import React from 'react'
import { Provider } from 'react-redux'
import { initStore, initTypeorm, ThunkDispatch } from './store'
import { navigationActions } from 'src/actions'
import { StatusBar, View } from 'react-native'
import { RoutesContainer } from './routes'
import { AppLoadingAndNotifications } from './ui/generic/appLoadingAndNotifications'
import { useScreens } from 'react-native-screens'
import { isNil } from 'ramda'
import {
  NavigationContainerComponent,
  NavigationRoute,
  NavigationState,
} from 'react-navigation'
import { setActiveNotificationFilter } from './actions/notifications'
import { black } from './styles/colors'
import { LoadingSpinner } from './ui/generic'

import {
  JolocomLinking,
  JolocomWebSockets,
  JolocomKeychainPasswordStore,
  JolocomSDK,
} from 'react-native-jolocom'

useScreens()

/**
 * NOTE: this is *not* exported on purpose
 * Other parts of the app should generally *not* access the store directly, but
 * rather through being connect()ed through redux
 *
 * If you think you need to export this, then something else probably needs
 * better architecture.
 */
let store: ReturnType<typeof initStore>
let sdkPromise: Promise<JolocomSDK>

export default class App extends React.PureComponent<
  {},
  { ready: boolean, showStatusBar: boolean }
> {
  private navigator!: NavigationContainerComponent

  public constructor(props: {}) {
    super(props)
    this.state = {
      ready: false,
      showStatusBar: true,
    }
    // only init store once, or else Provider complains (especially on 'toggle
    // inspector')
    //
    // but it needs to be done only when a new App is
    // instantiated because otherwise the overrides at the top of index.ts will
    // have not been excuted yet (while files are being imported) and initStore
    // triggers creation of BackendMiddleware which needs those
    if (!sdkPromise) {
      sdkPromise = initTypeorm().then(async storage => {
        const passwordStore = new JolocomKeychainPasswordStore()
        const sdk = new JolocomSDK({ storage, passwordStore })
        await sdk.usePlugins(new JolocomLinking(), new JolocomWebSockets())
        sdk.setDefaultDidMethod('jun')

        store = initStore(sdk)
        this.setState({ ready: true })
        return sdk
      })
    }
  }

  public handleNavigationChange(
    prevState: NavigationState,
    newState: NavigationState,
  ) {
    // @ts-ignore
    let navigation = this.navigator._navigation
    let curState: NavigationState | NavigationRoute = newState,
      navigationOptions

    while (curState.routes) {
      curState = curState.routes[curState.index]
      const childNav = navigation.getChildNavigation(curState.key)
      navigationOptions = navigation.router.getScreenOptions(childNav)
      navigation = childNav
    }

    const { notifications, statusBar } = navigationOptions

    if (!isNil(notifications)) {
      const thunkDispatch: ThunkDispatch = store.dispatch
      thunkDispatch(setActiveNotificationFilter(notifications))
    }

    if (!isNil(statusBar)) {
      this.setState({ showStatusBar: statusBar })
    } else {
      this.setState({ showStatusBar: true })
    }
  }

  private setNavigator(nav: NavigationContainerComponent | null) {
    if (!nav) return
    this.navigator = nav
    navigationActions.setTopLevelNavigator(this.navigator)
  }

  public render() {
    const { showStatusBar, ready } = this.state

    return (
      <React.Fragment>
        <StatusBar hidden={!showStatusBar} translucent />
        {showStatusBar && (
          <View
            style={{
              width: '100%',
              height: StatusBar.currentHeight,
              backgroundColor: black,
            }}
          />
        )}
        {!ready ? <LoadingSpinner /> :
          <Provider store={store}>
            <View style={{ flex: 1 }}>
              <RoutesContainer
                onNavigationStateChange={this.handleNavigationChange.bind(this)}
                ref={nav => this.setNavigator(nav)}
              />
              <AppLoadingAndNotifications />
            </View>
          </Provider>
        }
      </React.Fragment>
    )
  }
}
