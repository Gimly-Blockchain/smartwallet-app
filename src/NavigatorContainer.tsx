import React from 'react'
import { connect } from 'react-redux'
import { BackHandler, Linking, StatusBar } from 'react-native'
import { RootState } from 'src/reducers/'
import { navigationActions, accountActions, genericActions } from 'src/actions/'
import { routeList } from './routeList'
import { LoadingSpinner } from 'src/ui/generic/loadingSpinner'
import { ThunkDispatch } from './store'
import { handleDeepLink } from './actions/navigation'
import { toggleLoading } from './actions/account'
import { RoutesContainer } from './routes'
import { withErrorHandling, withLoading } from './actions/modifiers'
import { showErrorScreen } from './actions/generic'

import { useScreens } from 'react-native-screens'
import { NavigationActions } from 'react-navigation'
useScreens()

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps> {}

const darkBackgroundPages: string[] = [
  routeList.Landing,
  routeList.SeedPhrase,
  routeList.Exception,
  routeList.Loading,
  routeList.Entropy,
]

export class NavigatorContainer extends React.Component<Props> {
  private navigator: any

  constructor(props: Props) {
    super(props)
  }

  async componentDidMount() {
    await this.props.initApp()
    await this.props.checkIfAccountExists()
    const url = await Linking.getInitialURL()
    if (url) {
      this.props.handleDeepLink(url)
    }

    Linking.addEventListener('url', this.handleOpenURL)
    BackHandler.addEventListener('hardwareBackPress', this.navigateBack)
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.navigateBack)
    Linking.removeEventListener('url', this.handleOpenURL)
  }

  private navigateBack = () => {
    // return false if app exit is desired
    if (!this.navigator) return
    const { index, routes } = this.navigator.state.nav
    if (index === 0 && routes.length === 1 && routes[0].index === 0) {
      return false
    }

    this.navigator.dispatch(NavigationActions.back())

    return true
  }

  // When handleOpenURL is called, we pass the event url to the navigate method.
  private handleOpenURL = (event: any) => {
    this.props.handleDeepLink(event.url)
  }

  private setNavigator(nav: any) {
    if (!nav) return
    this.navigator = nav
    navigationActions.setTopLevelNavigator(this.navigator)
  }

  render() {
    let isDarkBackground = false
    if (this.navigator) {
      const { routes, index } = this.navigator.state.nav
      const currentRoute = routes[index].routeName
      isDarkBackground = darkBackgroundPages.includes(currentRoute)
    }

    return (
      <React.Fragment>
        <StatusBar barStyle={isDarkBackground ? 'light-content' : 'default'} />
        <RoutesContainer ref={nav => this.setNavigator(nav)} />
        {this.props.deepLinkLoading && <LoadingSpinner />}
      </React.Fragment>
    )
  }
}

const mapStateToProps = (state: RootState) => ({
  deepLinkLoading: state.sso.deepLinkLoading,
})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  handleDeepLink: async (url: string) =>
    dispatch(
      withLoading(toggleLoading)(
        withErrorHandling(showErrorScreen)(handleDeepLink(url)),
      ),
    ),
  checkIfAccountExists: () =>
    dispatch(
      withLoading(toggleLoading)(
        withErrorHandling(showErrorScreen)(accountActions.checkIdentityExists),
      ),
    ),
  initApp: () => dispatch(genericActions.initApp),
})
export const Navigator = connect(
  mapStateToProps,
  mapDispatchToProps,
)(NavigatorContainer)
