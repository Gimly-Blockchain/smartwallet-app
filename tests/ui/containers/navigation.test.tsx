import React from 'react'
import { NavigatorContainer } from 'src/NavigatorContainer'
import { shallow } from 'enzyme'
import { routeList } from 'src/routeList'

describe('NavigatorContainer', () => {
  const COMMON_PROPS: Partial<NavigatorContainer['props']> = {
    goBack: jest.fn(),
    deepLinkLoading: false,
    initApp: jest.fn(),
    handleDeepLink: jest.fn(),
    checkIfAccountExists: jest.fn(),
  }

  it('mounts correctly and matches snapshot', () => {
    const navigation = {
      index: 0,
      routes: [{ routeName: 'Home' }],
      navigate: jest.fn(),
    }

    const props: NavigatorContainer['props'] = {
      ...COMMON_PROPS,
      // @ts-ignore
      navigation
    }

    const rendered = shallow(<NavigatorContainer { ...props } />)
    expect(rendered).toMatchSnapshot()
  })

  it('correctly hides bottom action bar', () => {
    const navigation = {
      index: 0,
      routes: [{ routeName: routeList.Landing}],
      navigate: jest.fn(),
    }

    const landingPageProps: NavigatorContainer['props'] = {
      ...COMMON_PROPS,
      // @ts-ignore
      navigation
    }

    const rendered = shallow(<NavigatorContainer {...landingPageProps} />)
    expect(rendered).toMatchSnapshot()
  })
})
