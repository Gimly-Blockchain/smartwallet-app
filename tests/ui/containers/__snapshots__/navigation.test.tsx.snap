// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigatorContainer correctly hides bottom action bar 1`] = `
<React.Fragment>
  <StatusBar
    animated={false}
    barStyle="light-content"
    showHideTransition="fade"
  />
  <withOrientation
    style={
      Object {
        "backgroundColor": "black",
        "flex": 1,
      }
    }
  >
    <NavigationContainer
      navigation={
        Object {
          "addListener": [Function],
          "dispatch": [MockFunction],
          "getParam": [Function],
          "goBack": [Function],
          "navigate": [Function],
          "pop": [Function],
          "popToTop": [Function],
          "push": [Function],
          "replace": [Function],
          "setParams": [Function],
          "state": Object {
            "index": 0,
            "navigate": [MockFunction],
            "routes": Array [
              Object {
                "routeName": "Landing",
              },
            ],
          },
        }
      }
    />
  </withOrientation>
</React.Fragment>
`;

exports[`NavigatorContainer mounts correctly and matches snapshot 1`] = `
<React.Fragment>
  <StatusBar
    animated={false}
    barStyle="dark-content"
    showHideTransition="fade"
  />
  <withOrientation
    style={
      Object {
        "backgroundColor": "white",
        "flex": 1,
      }
    }
  >
    <NavigationContainer
      navigation={
        Object {
          "addListener": [Function],
          "dispatch": [MockFunction],
          "getParam": [Function],
          "goBack": [Function],
          "navigate": [Function],
          "pop": [Function],
          "popToTop": [Function],
          "push": [Function],
          "replace": [Function],
          "setParams": [Function],
          "state": Object {
            "index": 0,
            "navigate": [MockFunction],
            "routes": Array [
              Object {
                "routeName": "Home",
              },
            ],
          },
        }
      }
    />
  </withOrientation>
  <BottomActionBar
    openScanner={[MockFunction]}
  />
</React.Fragment>
`;
